test/src/module/common/typePackTest.c

```diff
@@ -0,0 +1,315 @@
+                "988401"                                            //      6, time, 66
+                "00"                                                //     array end
+            "2806050403020100"                                      // 49,  bin, 0x050403020100
+            "21"                                                    // 51,  bin, zero length
+            "00",                                                   // end
+            "check pack hex");
+
+        // -------------------------------------------------------------------------------------------------------------------------
+        TEST_TITLE("read pack");
+
+        IoRead *read = ioBufferReadNew(pack);
+        ioReadOpen(read);
+
+        PackRead *packRead = NULL;
+        TEST_ASSIGN(packRead, pckReadNew(read), "new read");
+
+        TEST_RESULT_UINT(pckReadU64P(packRead), 0750, "read mode");
+        TEST_RESULT_UINT(pckReadU64P(packRead), 1911246845, "read timestamp");
+        TEST_ERROR(pckReadU64P(packRead, .id = 2), FormatError, "field 2 was already read");
+        TEST_ERROR(pckReadU32P(packRead, .id = 7), FormatError, "field 7 is type 'u64' but expected 'u32'");
+        TEST_RESULT_UINT(pckReadU64P(packRead, .id = 7), 0xFFFFFFFFFFFFFFFF, "read max u64");
+        TEST_ERROR(pckReadU64P(packRead, .id = 9), FormatError, "field 9 does not exist");
+        TEST_RESULT_BOOL(pckReadNullP(packRead, .id = 9), true, "field 9 is null");
+        TEST_RESULT_BOOL(pckReadNullP(packRead, .id = 10), false, "field 10 is not null");
+        TEST_RESULT_UINT(pckReadU64P(packRead, .id = 10), 1, "read 1");
+        TEST_RESULT_UINT(pckReadU32P(packRead, .id = 12, .defaultNull = true), 127, "read 127 (skip field 11)");
+        TEST_RESULT_INT(pckReadI64P(packRead), -1, "read -1");
+        TEST_RESULT_INT(pckReadI32P(packRead, .id = 14), -1, "read -1");
+        TEST_RESULT_BOOL(pckReadBoolP(packRead, .id = 15), true, "read true");
+        TEST_RESULT_BOOL(pckReadBoolP(packRead, .id = 20), false, "read false");
+
+        TEST_ERROR(pckReadObjEndP(packRead), FormatError, "not in object");
+        TEST_RESULT_VOID(pckReadObjBeginP(packRead, .id = 28), "read object begin");
+        TEST_ERROR(pckReadArrayEndP(packRead), FormatError, "not in array");
+        TEST_RESULT_BOOL(pckReadBoolP(packRead, .defaultNull = true), true, "read true");
+        TEST_RESULT_BOOL(pckReadBoolP(packRead), false, "read false");
+        TEST_ERROR(pckReadBoolP(packRead), FormatError, "field 3 does not exist");
+        TEST_RESULT_BOOL(pckReadNullP(packRead), true, "field 3 is null");
+        TEST_RESULT_BOOL(pckReadNullP(packRead, .id = 4), true, "field 4 is null");
+        TEST_RESULT_BOOL(pckReadBoolP(packRead, .defaultNull = true), false, "read default false");
+        TEST_RESULT_VOID(pckReadObjEndP(packRead), "read object end");
+
+        TEST_ERROR(pckReadArrayEndP(packRead), FormatError, "not in array");
```

Comment:
pckReadNext() skips NULLs so this will not work the way you expected. Basically, pckReadNext() is going to read the next thing that *exists*.

Reply:
Hmmm, well that does not explain the behavior since 29 through 36 do not exist so it should have been "skipping" and still doesn't explain the error when reading the non-existent 31.

Reply:
So, the main thing is the code doesn't pay attention to the type of the next field (which is an array) and just tries to keep reading integers. The first one works OK because all there are all NULLs up to the array. So the defaultNull works as expected.

But the next pckReadNext() reads the first u64 of the array, which happens to be zero so it looks like the NULL did before. But then the next read is to a u64 that is 1 (the second element of the array) so the code dies.

So the test code here is not right, but there's also no error so it's not at all clear what has gone wrong.  So I've added an error that if the next type is a container and you don't read that container it will error. In other words, you can't just "read past" a container because it's not clear if you want the data in the container or the data after the container.

Reply:
Fix in 4f1302fd9c67de86eea4b1ed5a76217759eccbc0 looks good.